mgmake/generate.go

```diff
@@ -1,94 +1,168 @@
 package mgmake
 
 import (
+	"bytes"
 	"fmt"
 	"os"
 	"path/filepath"
+	"reflect"
+	"runtime"
 	"strings"
 	"text/template"
 
 	"github.com/iancoleman/strcase"
-	"github.com/magefile/mage/sh"
-	"go.einride.tech/mage-tools/mglog"
+	"go.einride.tech/mage-tools/mgtool"
+	"go.einride.tech/mage-tools/mgtool/mgpath"
 )
 
 type templateTargets struct {
 	MakeTarget string
 	MageTarget string
 	Args       []string
+	LastIndex  bool
 }
 
-// GenerateMakefile is a mage target that ...
-func GenerateMakefile(makefile string) error {
-	const defaultTargets = "mage_targets"
-	genDir := filepath.Dir(makefile)
-	mgDefaultTargets := fmt.Sprintf("%s.mk", filepath.Join(genDir, strcase.ToKebab(defaultTargets)))
-	if makefile == mgDefaultTargets {
-		return fmt.Errorf("%s has the same name as the default %s makefile", makefile, mgDefaultTargets)
+// nolint: gochecknoglobals
+var (
+	makefiles = make(map[string]makefile)
+)
+
+type makefile struct {
+	Path          string
+	DefaultTarget string
+}
+
+type Makefile struct {
+	Namespace     interface{}
+	Path          string
+	DefaultTarget func()
+}
+
+func GenerateMakefiles(mks ...Makefile) {
+	for _, i := range mks {
+		if i.Path == "" {
+			panic("Path needs to be defined")
+		}
+		namespace := defaultNamespace
+		if i.Namespace != nil {
+			namespace = reflect.TypeOf(i.Namespace).Name()
+		}
+		defaultTarget := runtime.FuncForPC(reflect.ValueOf(i.DefaultTarget).Pointer()).Name()
+		defaultTarget = strings.TrimPrefix(defaultTarget, "main.")
+		defaultTarget = strings.TrimPrefix(defaultTarget, namespace+".")
+		defaultTarget = strings.Split(defaultTarget, "-")[0]
+		makefiles[toMakeTarget(namespace)] = makefile{Path: i.Path, DefaultTarget: defaultTarget}
+	}
+}
+
+const defaultNamespace = "default"
+
+func GenMakefiles() error {
+	if len(makefiles) == 0 {
+		return fmt.Errorf("no makefiles to generate, " +
+			"provide some info about adding an init method or linking to readme")
```

Comment:
We need a better message here :P

Reply:
not specifying any makefiles to generate won't trigger this error, right?

Reply:
If you try and run `GenMakefiles` without having an `init` method where you added makefiles in the magefile, this error will trigger